src/shogun/machine/GLM.h

```diff
@@ -0,0 +1,296 @@
+		 * stop when relative change in parameter norm is below the threshold.
+		 */
+		float64_t m_tolerance = 1e-6;
+
+		float64_t m_learning_rate = 2e-1;
+
+		bool m_compute_bias = true;
+
+		std::shared_ptr<GradientDescendUpdater> m_gradient_updater;
+
+		// std::shared_ptr<ConstLearningRate> m_learning_rate;
+
+		std::shared_ptr<ElasticNetPenalty> m_penalty;
+
+		std::shared_ptr<GLMCostFunction> m_cost_function;
+	};
+
+	class GLMCostFunction
+	{
+	public:
+		friend class GLM;
+
+		GLMCostFunction()
+		{
+		}
+
+		virtual SGVector<float64_t> get_gradient_weights(
+		    const SGMatrix<float64_t> X, const SGVector<float64_t> y,
+		    const SGVector<float64_t> w, const float64_t bias,
+		    const float64_t lambda, const float64_t alpha,
+		    const bool compute_bias, const float64_t eta,
+		    const GLM_DISTRIBUTION distribution)
+		{
+			auto n_samples = y.vlen;
+
+			auto z = compute_z(X, w, bias);
+			auto mu = non_linearity(z, compute_bias, eta, distribution);
+			auto grad_mu = gradient_non_linearity(z, eta, distribution);
+
+			SGVector<float64_t> grad_w(w.vlen);
+			SGVector<float64_t> a;
+			switch (distribution)
+			{
+			case POISSON: {
+				// grad_w = ((grad_mu.T)⚬X - ((y*grad_mu/mu).T)⚬X).T
+				a = linalg::element_prod(y, grad_mu);
+				a = linalg::element_div(a, mu);
+
+				auto prod1 =
+				    linalg::matrix_prod(SGMatrix(grad_mu), X, true, true);
+				auto prod2 = linalg::matrix_prod(SGMatrix(a), X, true, true);
+
+				grad_w = linalg::transpose_matrix(
+				    linalg::add(prod1, prod2, 1.0, -1.0));
+
+				break;
+			}
+			default:
+				error(
+				    "Distribution type {} not implemented.",
+				    GLM::glm_enum_to_string(distribution));
+				break;
+			}
+
+			grad_w = linalg::scale(grad_w, 1.0 / n_samples);
+			grad_w = linalg::add(grad_w, w, 1.0, lambda * (1 - alpha));
+
+			return grad_w;
+		}
+
+		virtual float64_t get_gradient_bias(
+		    const SGMatrix<float64_t> X, const SGVector<float64_t> y,
+		    const SGVector<float64_t> w, const float64_t bias,
+		    const bool compute_bias, const float64_t eta,
+		    const GLM_DISTRIBUTION distribution)
+		{
+			auto n_samples = y.vlen;
+			auto z = compute_z(X, w, bias);
+			auto mu = non_linearity(z, compute_bias, eta, distribution);
+			auto grad_mu = gradient_non_linearity(z, eta, distribution);
+
+			float64_t grad_bias = 0;
+			switch (distribution)
+			{
+			case POISSON:
+				grad_bias += SGVector<float64_t>::sum(grad_mu);
+				grad_mu = linalg::element_prod(y, grad_mu);
+				grad_mu = linalg::element_div(grad_mu, mu);
+				grad_bias -= SGVector<float64_t>::sum(grad_mu);
+				break;
+
+			default:
+				error(
+				    "Distribution type {} not implemented.",
+				    GLM::glm_enum_to_string(distribution));
+				break;
+			}
+			grad_bias /= n_samples;
+
+			return grad_bias;
+		}
+
+		const char* get_name() const
+		{
+			return "GLMCostFunction";
+		}
+
+	private:
+		virtual const SGVector<float64_t> compute_z(
```

Comment:
why virtual?

Reply:
Only the functions which will be reimlemented should be virtual. right.

Reply:
There is no inheritance here right? Or would you expect a developer to inherit from `GLMCostFunction`?

Reply:
Yes, that is a plan for the future: When one wants to implement something else like probit or gamma one will have to inherit from this and make a class called GLMGammaCostFunction. This is to prevent all the switch case clutter.

Reply:
Oh I see, that makes sense. But the linear term is always computed as `Xw + b`?

Reply:
Yes.

Reply:
if it is always the same it doesn't make sense for this to be virtual then?

Reply:
Yes, I have made is non-virtual :)